### PR TITLE
layer-level fetching scheme

### LINKED ISSUE #109: Add interaction type to API call
In order to generate correct statistics, we need to know whether a user interaction is a pan or zoom. Since the user interaction is handled on the front end, the best way to do this is to determine when there is a jump, and add this information as a isZooming boolean variable in the API call to the backend. This issue is related to the data collection for fetch times issue:
https://github.com/tracyhenry/Kyrix/issues/74

### LINKED ISSUE #111: make deltabox and fetching scheme per-layer
Dynamic boxes are good for most types of layers, but some layers are better with tiles, or dynamic boxes without delta fetching. 

Instead of being a per-project config parameter, Dbox/tile should be configurable on a per-layer basis (default to dbox)

### PR DESCRIPTION
**Pull request by tracyhenry**
Fetching scheme can now be configured for each layer in the compiler, for example:
```
layer.setFetchingScheme("tiling")      // tiling
layer.setFetchingScheme("dbox")        // dynamic box w/ delta boxes, which is also default
layer.setFetchingScheme("dbox", false) // dynamic box w/o delta boxes
```

 It was a global parameter in `parameter.js` 

fixes #111 
fixes #109

### CODE REVIEW

**Comment by tracyhenry** [comment]
@peterg17 I did a fix to #109: https://github.com/tracyhenry/Kyrix/pull/112/commits/7f1d071421bd4abdf1cc1a30f1cff5d643980db2

You can also help review other changes to the backend, which I tried best to keep minimal...

**Comment by tracyhenry** [comment]
@houxinli can you also help review the changes, thanks!

**Comment by houxinli** [comment]
Sure thing

________________________________
发件人: Wenbo Tao <notifications@github.com>
发送时间: Friday, August 30, 2019 2:11:21 AM
收件人: tracyhenry/Kyrix <Kyrix@noreply.github.com>
抄送: Xinli Hou <houxinli@hotmail.com>; Mention <mention@noreply.github.com>
主题: Re: [tracyhenry/Kyrix] layer-level fetching scheme (#112)


@houxinli<https://github.com/houxinli> can you also help review the changes, thanks!

―

**Comment by peterg17** [comment]
Thank you for adding this, i'll review the backend code

File: back-end/src/main/java/box/BoxGetter.java [hunk 1]
```diff
@@ -11,80 +11,40 @@
...
-                                .getDataFromRegion(
-                                        c, i, deltaWkt, predicates.get(i), newBox, oldBox));
+            if (curLayer.isStatic() || !curLayer.getFetchingScheme().equals("dbox")) {
+                data.add(new ArrayList<>());
+                continue;
+            }
+            // calculate delta area
+            Box curOldBox;
+            if (curLayer.isDeltaBox()) curOldBox = oldBox;
+            else curOldBox = DUMMY_OLD_BOX;
```
Comment (peterg17) [inline back-end/src/main/java/box/BoxGetter.java]: is this dummy box only used upon initialization when there is no old box?

**Comment by peterg17** [comment]
Ok I checked all of the backend code and couldn't find anything that immediately stood out to me. I also tested locally to test the isPanning variable. One thing in terms of using the new code is that It seemed to work for zoom-ins, but set isPanning=true on zoom-outs. Do we want to register zoom-outs as zooms? I assume so

**Comment by tracyhenry** [comment]
> Ok I checked all of the backend code and couldn't find anything that immediately stood out to me. I also tested locally to test the isPanning variable. One thing in terms of using the new code is that It seemed to work for zoom-ins, but set isPanning=true on zoom-outs. Do we want to register zoom-outs as zooms? I assume so

setting `isPanning` to true on zoom-outs should be a bug. I'll investigate and get back to you

File: back-end/src/main/java/box/BoxGetter.java (hunk 1 above)
Comment (tracyhenry) [reply back-end/src/main/java/box/BoxGetter.java]: Actually, no. 

This `DUMMY_OLD_BOX` is used in the backend for dynamic boxes with no delta box, in which case the backend will ignore the old box sent by the front-end. 

Note that previously the front-end will send an old box if the `deltaBox` is configure to be globally true. Now that it's configured per layer, I let the front-end always send an old box (will be dummy values when there is no old box), and let the back-end use it for layers that require `deltaBox`. So for layers that do not require `deltaBox`, the backend will use this `DUMMY_OLD_BOX` and ignore the old box sent by the front-end.

**Comment by tracyhenry** [comment]
> Ok I checked all of the backend code and couldn't find anything that immediately stood out to me. I also tested locally to test the isPanning variable. One thing in terms of using the new code is that It seemed to work for zoom-ins, but set isPanning=true on zoom-outs. Do we want to register zoom-outs as zooms? I assume so

Peter, the test I did showed that `isPanning` is indeed set to true on some zoom-outs. However, it's only set to true when the zoom out is happening within a canvas (let me know if it's not the case in your tests). 

So I think I actually implemented `! isJumping`..... Yet `isJumping` should be the thing you are looking for, right? In the performance test, I think we only need to care about whether a request happens right after a canvas switch. 

Let me know if this makes sense. I can go ahead and change it to `isJumping`.

File: front-end/js/dynamicLayers.js [hunk 2]
```diff
@@ -107,60 +107,64 @@ function highlightLowestSvg(viewId, svg, layerId) {
...
+                gvd.renderData[i] = gvd.renderData[i].filter(function(d) {
+                    return mp.hasOwnProperty(JSON.stringify(d))
+                        ? false
+                        : (mp[JSON.stringify(d)] = true);
                 });
-            tiles.exit().remove();
-        });
+
+                // remove exit (invisible) tiles
+                d3.select(".layerg.layer" + i)
```
Comment (houxinli) [inline front-end/js/dynamicLayers.js]: I'm not sure if this is ok if there are multiple views?

File: compiler/src/Layer.js [hunk 3]
```diff
@@ -18,6 +18,8 @@ function Layer(transform, isStatic) {
     this.transform = transform;
     if (isStatic == null) this.isStatic = false;
     else this.isStatic = isStatic;
+    this.fetchingScheme = "dbox";
+    this.deltaBox = true;
     this.isAutoDDLayer = false;
     this.isPredicatedTable = false;
```
[the remark this answers is not in the document]
Comment (tracyhenry) [reply compiler/src/Layer.js]: yes we talked about this

File: front-end/js/dynamicLayers.js (hunk 2 above)
Comment (tracyhenry) [reply front-end/js/dynamicLayers.js]: good point, this should be a bug

**Comment by peterg17** [comment]
@wenbo, I tested your isPanning variable with the dots-uniform example, by zooming in and out and then printing out the value of isPanning. It was set to true on zoom outs, which I think is a switch between the lower level canvas and the upper level canvas, so it is a canvas switch. Is that accurate?

**Comment by tracyhenry** [comment]
> wenbo, I tested your isPanning variable with the dots-uniform example, by zooming in and out and then printing out the value of isPanning. It was set to true on zoom outs, which I think is a switch between the lower level canvas and the upper level canvas, so it is a canvas switch. Is that accurate?

can you paste here the back-end log when you ran the test? The log should have isPanning printed.

**Pull request merged**